File: apps/web/src/lib/ai/vector-store.ts
import { prisma } from "@hiresense/db";

export async function searchFreelancers(params: {
  queryEmbedding: number[];
  limit?: number;
  category?: string;
  city?: string;
  minRating?: number;
}) {
  const { queryEmbedding, limit = 10, category, city, minRating } = params;
  const embeddingStr = `[${queryEmbedding.join(",")}]`;

  const conditions: string[] = [];
  const values: unknown[] = [];
  let paramIndex = 1;

  if (category) {
    conditions.push(`fp.category = $${paramIndex++}`);
    values.push(category);
  }
  if (city) {
    conditions.push(`fp.city = $${paramIndex++}`);
    values.push(city);
  }
  if (minRating) {
    conditions.push(`fp."avgRating" >= $${paramIndex++}`);
    values.push(minRating);
  }

  const whereClause = conditions.length > 0 ? `WHERE ${conditions.join(" AND ")}` : "";

  const results = await prisma.$queryRawUnsafe(
    `SELECT
      fp.id,
      fp."userId",
      fp."displayName",
      fp.headline,
      fp.bio,
      fp.city,
      fp.state,
      fp."hourlyRate",
      fp."avgRating",
      fp."completedGigs",
      fp.embedding <=> '${embeddingStr}'::vector AS distance
    FROM "FreelancerProfile" fp
    ${whereClause}
    ORDER BY distance
    LIMIT $${paramIndex}`,
    ...values,
    limit
  );

  return results;
}

export async function searchGigs(params: {
  queryEmbedding: number[];
  limit?: number;
  category?: string;
}) {
  const { queryEmbedding, limit = 10, category } = params;
  const embeddingStr = `[${queryEmbedding.join(",")}]`;

  const conditions: string[] = [`g.status = 'OPEN'`];
  const values: unknown[] = [];
  let paramIndex = 1;

  if (category) {
    conditions.push(`g.category = $${paramIndex++}`);
    values.push(category);
  }

  const whereClause = `WHERE ${conditions.join(" AND ")}`;

  const results = await prisma.$queryRawUnsafe(
    `SELECT
      g.id,
      g.title,
      g.description,
      g.category,
      g."budgetMin",
      g."budgetMax",
      g."isRemote",
      g.city,
      g.embedding <=> '${embeddingStr}'::vector AS distance
    FROM "Gig" g
    ${whereClause}
    ORDER BY distance
    LIMIT $${paramIndex}`,
    ...values,
    limit
  );

  return results;
}

File: apps/ai-services/app/services/vector_store.py
import psycopg2
from app.core.config import settings


def get_connection():
    """Get a connection to Supabase PostgreSQL."""
    return psycopg2.connect(settings.database_url)


def similarity_search_freelancers(
    query_embedding: list[float],
    limit: int = 10,
    category: str | None = None,
    city: str | None = None,
    min_rating: float | None = None,
) -> list[dict]:
    """Search freelancer profiles by vector similarity using pgvector."""
    conn = get_connection()
    try:
        cur = conn.cursor()

        # Build query with optional filters
        conditions = []
        params = []

        if category:
            conditions.append("fp.category = %s")
            params.append(category)
        if city:
            conditions.append("fp.city = %s")
            params.append(city)
        if min_rating:
            conditions.append('fp."avgRating" >= %s')
            params.append(min_rating)

        where_clause = ""
        if conditions:
            where_clause = "WHERE " + " AND ".join(conditions)

        embedding_str = "[" + ",".join(str(x) for x in query_embedding) + "]"

        query = f"""
            SELECT
                fp.id,
                fp."userId",
                fp."displayName",
                fp.headline,
                fp.bio,
                fp.city,
                fp.state,
                fp."hourlyRate",
                fp."avgRating",
                fp."completedGigs",
                fp.embedding <=> '{embedding_str}'::vector AS distance
            FROM "FreelancerProfile" fp
            {where_clause}
            ORDER BY distance
            LIMIT %s
        """
        params.append(limit)

        cur.execute(query, params)
        columns = [desc[0] for desc in cur.description]
        results = [dict(zip(columns, row)) for row in cur.fetchall()]
        return results
    finally:
        conn.close()


def similarity_search_gigs(
    query_embedding: list[float],
    limit: int = 10,
    category: str | None = None,
    is_remote: bool | None = None,
) -> list[dict]:
    """Search gigs by vector similarity using pgvector."""
    conn = get_connection()
    try:
        cur = conn.cursor()

        conditions = ['g.status = %s']
        params = ["OPEN"]

        if category:
            conditions.append("g.category = %s")
            params.append(category)
        if is_remote is not None:
            conditions.append('g."isRemote" = %s')
            params.append(is_remote)

        where_clause = "WHERE " + " AND ".join(conditions)
        embedding_str = "[" + ",".join(str(x) for x in query_embedding) + "]"

        query = f"""
            SELECT
                g.id,
                g.title,
                g.description,
                g.category,
                g."budgetMin",
                g."budgetMax",
                g."isRemote",
                g.city,
                g.embedding <=> '{embedding_str}'::vector AS distance
            FROM "Gig" g
            {where_clause}
            ORDER BY distance
            LIMIT %s
        """
        params.append(limit)

        cur.execute(query, params)
        columns = [desc[0] for desc in cur.description]
        results = [dict(zip(columns, row)) for row in cur.fetchall()]
        return results
    finally:
        conn.close()
